### PR TITLE
Train add

### PR DESCRIPTION
**Pull request by StephenHausler**

### CODE REVIEW

File: patchnetvlad/configs/train.ini [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+momentum = 0.9
+patience = 10
+evalevery = 1
+margin = 0.1
+num_clusters = 16
+includepanos = False
+imageresizew = 640
+imageresizeh = 480
+
+[val]
```
Comment (Tobias-Fischer) [inline patchnetvlad/configs/train.ini:19]: Where are those values being used? (all the ones in the `[val]` group)

File: patchnetvlad/tools/msls.py [hunk 2]
```diff
@@ -0,0 +1,549 @@
+import numpy as np
```
Comment (Tobias-Fischer) [inline patchnetvlad/tools/msls.py:1]: This file requires a license header and reference to https://github.com/mapillary/mapillary_sls on which the file is heavily based on.

Comment (Tobias-Fischer) [reply patchnetvlad/tools/msls.py:1]: (best to copy their license header, and add a "modified by ..")

File: train.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+            opt.start_epoch = checkpoint['epoch']
+
+            print("=> loaded checkpoint '{}'".format(opt.resume_path, ))
+        else:
+            raise FileNotFoundError("=> no checkpoint found at '{}'".format(opt.resume_path))
+    else: # if not, assume fresh training instance and will initially generate cluster centroids
+        print('===> Finding cluster centroids')
+
+        print('===> Loading dataset(s) for clustering')
+        train_dataset = MSLS(opt.dataset_root_dir, mode='test', cities='train', transform=input_transform(),
```
Comment (Tobias-Fischer) [inline train.py:367]: I think this is only needed in the `else` part of the `if opt.cluster_path:` statement?

File: train.py [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+    # write checkpoints in logdir
+    logdir = writer.file_writer.get_logdir()
+    opt.save_file_path = join(logdir, 'checkpoints')
+    makedirs(opt.save_file_path)
+
+    not_improved = 0
+    best_score = 0
+    if opt.resume_path:
+        # noinspection PyUnboundLocalVariable
+        if 'not_improved' in checkpoint:
```
Comment (Tobias-Fischer) [inline train.py:454]: I think this can be removed - `not_improved` should always be in the checkpoint now.

**Review by Tobias-Fischer — CHANGES_REQUESTED** [review]
Looks good overall, thanks @StephenHausler.

I have a few minor requests (see inline comments) as well as two bigger ones:

1) It would be great to move the `train`, `test` and `get_clusters` methods to separate files living in `patchnetvlad/training_tools`. Similarly, move `save_checkpoint` and `humanbytes` into `patchnetvlad/training_tools/tools.py`.

2) The readme needs extending with describing how the training can be performed.

**Comment by Tobias-Fischer** [comment]
Fix #8

File: patchnetvlad/tools/msls.py (hunk 2 above)
Comment (StephenHausler) [reply patchnetvlad/tools/msls.py:1]: Done

File: patchnetvlad/configs/train.ini (hunk 1 above)
Comment (StephenHausler) [reply patchnetvlad/configs/train.ini:19]: Actually they are not being used anywhere, I'll delete the [val] group

File: train.py (hunk 3 above)
Comment (StephenHausler) [reply train.py:367]: Good point, I moved it

File: train.py (hunk 4 above)
Comment (StephenHausler) [reply train.py:454]: Yep, I think so too

File: patchnetvlad/training_tools/msls.py [hunk 5]
```diff
@@ -42,11 +42,11 @@
...
 
-
+# 'train': ["trondheim", "london", "boston", "melbourne", "amsterdam", "helsinki",
+#           "tokyo", "toronto", "saopaulo", "moscow", "zurich", "paris", "bangkok",
+#           "budapest", "austin", "berlin", "ottawa", "phoenix", "goa", "amman", "nairobi", "manila"],
 default_cities = {
-    'train': ["trondheim", "london", "boston", "melbourne", "amsterdam", "helsinki",
-           "tokyo", "toronto", "saopaulo", "moscow", "zurich", "paris", "bangkok",
-           "budapest", "austin", "berlin", "ottawa", "phoenix", "goa", "amman", "nairobi", "manila"],
+    'train': ["trondheim"],
```
Comment (Tobias-Fischer) [inline patchnetvlad/training_tools/msls.py:49]: This should be changed back to all cities.

**Review by Tobias-Fischer — CHANGES_REQUESTED** [review]
Looks good to me now, thanks! One very minor comment below (probably left from testing) We just need the PCA functionality and this is ready to go in my opinion.

@oravus - would be great if you could have a look, too.

**Comment by StephenHausler** [comment]
All review comments completed, code has been tested and should be ready for merge.

**Comment by Tobias-Fischer** [comment]
Thanks a lot @StephenHausler - let's merge it in :)

**Pull request merged**